### PR TITLE
Cannot force allocate primary to a node where the shard already exists

### PR DESCRIPTION
**Pull request by abeyad**
Before, it was possible that the SameShardAllocationDecider would allow
force allocation of an unassigned primary to the same node on which an
active replica is assigned.  This could only happen with shadow replica
indices, because when a shadow replica primary fails, the replica gets
promoted to primary but in the INITIALIZED state, not in the STARTED
state (because the engine has specific reinitialization that must take
place in the case of shadow replicas).  Therefore, if the now promoted
primary that is initializing fails also, the primary will be in the
unassigned state, because replica to primary promotion only happens when
the failed shard was in the started state.  The now unassigned primary
shard will go through the allocation deciders, where the
SameShardsAllocationDecider would return a NO decision, but would still
permit force allocation on the primary if all deciders returned NO.

This commit implements canForceAllocatePrimary on the
SameShardAllocationDecider, which ensures that a primary cannot be
force allocated to the same node on which an active replica already
exists.

Relates #22021

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
Can you add a test in `SameShardRoutingTests`?

**Comment by abeyad** [comment]
@ywelsch I pushed a test in https://github.com/elastic/elasticsearch/pull/22031/commits/82c8137993844a699cde8886541a2f676bc3c323

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/SameShardRoutingTests.java [hunk 1]
```diff
@@ -86,4 +97,34 @@ public void testSameHost() {
...
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomIntBetween(2, 4), 1);
+        Index index = clusterState.getMetaData().index("idx").getIndex();
+        ShardRouting primaryShard = clusterState.routingTable().index(index).shard(0).primaryShard();
+        RoutingNode routingNode = clusterState.getRoutingNodes().node(primaryShard.currentNodeId());
+        RoutingAllocation routingAllocation = new RoutingAllocation(new AllocationDeciders(Settings.EMPTY, Collections.emptyList()),
+            new RoutingNodes(clusterState, false), clusterState, ClusterInfo.EMPTY, System.nanoTime(), false
+        );
+
+        // can't force allocate same shard copy to the same node
+        ShardRouting newPrimary = ShardRouting.newUnassigned(primaryShard.shardId(), true,
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/SameShardRoutingTests.java]: TestShardRouting automatically provides some of the randomness that you explicitly do here.

**Review by ywelsch — APPROVED** [review]
Left one suggestion. I think it's also ok if this doesn't go to 5.0.3

**Comment by abeyad** [comment]
I pushed 960304669859f725236ca820d93192233a86e071 which uses `TestShardRouting`.  Thanks for the review @ywelsch

**Comment by abeyad** [comment]
5.x commit: 6d3e6f2b9ccc4fc0fdb1fb3cd9bb2c22ab03e6e8

**Pull request merged**